### PR TITLE
[Feature] Introduce TenAnn as the vector index implementation library

### LINKED ISSUE #46678: [Feature] Support vector index and ANNS.
- [x] https://github.com/StarRocks/starrocks/pull/48521
- [x] https://github.com/StarRocks/starrocks/pull/49346
- [x] https://github.com/StarRocks/starrocks/pull/48384
- [x] https://github.com/StarRocks/starrocks/pull/47036

### **Is your feature request related to a problem? Please describe.**

**High-dimensional vector similarity search (HVSS)** has become a hot topic in recent years, due to its key role in content retrieval, recommendation systems, and **LLM RAG.** Popular database systems like ClickHouse, PostgreSQL, and TiDB have already implemented or are integrating HVSS. StarRocks has also introduced conventional brute-force HVSS solutions through functions like cosine_similarity. However, this approach suffers from significant efficiency and scalability challenges.

To enable efficient and scalable HVSS, adding **vector index** support into StarRocks' storage layer is essential. 
By leveraging **approximate nearest neighbor search (ANNS)** through vector indexes, we can achieve a significant enhancement in performance (potentially by several orders of magnitude), with a minor accuracy trade-off.

### **Describe the solution you'd like**

We propose to incorporate vector/ANNS indexes into StarRocks, including:
- **HNSW**
- **IVF-PQ**

For each index, we propose to support various distance/similarity metrics including:
- **Squared euclidean distance**: $| \boldsymbol{x} - \boldsymbol{y} |^2$
- **Cosine similarity**: $\frac{ \boldsymbol{x} \cdot  \boldsymbol{y} }{|\boldsymbol{x}||\boldsymbol{y}|}$

Moreover, we propose the utilization of vector indexes to accommodate a diverse range of queries:
- **ANNS**: retrieve the top-k most similar vectors to a given query.
- **Filtered ANNS** (also known as hybrid search): retrieve the top-k most similar vectors that satisfy specific structural predicates.
- **Range search**: retrieve all vectors within a specified distance threshold from the query vector.
- **Filtered range search**: retrieve all vectors satisfying specific structural predicates and within a specified distance threshold from the query vector.

### SQL Syntax 

Create a table with vector index:
```sql
CREATE TABLE `vector_table` (
  `id` varchar(100) NOT NULL COMMENT "",
  `pk` bigint(20) NULL COMMENT "",
  `score` float NOT NULL COMMENT "",
  `question` varchar(65535) NOT NULL COMMENT "",
  `answer` varchar(65535) NOT NULL COMMENT "",
  `source` varchar(65535) NOT NULL COMMENT "",
  `question_embedding` array<float> NOT NULL COMMENT "",
  INDEX my_index (`question_embedding`) USING VECTOR_INDEX PROPERTIES( 
    "dim" = "768",                            // vector dimension
    "index_type" = "hnsw" ,            // using HNSW index type
    "metric_type" = "l2_distance", // using squared euclidean distance
    "efSearch" = "40",                     // default search parameter for HNSW 
    "M" = "32",                                 // build parameter for HNSW
    "efConstruction" = "40",           // build parameter for HNSW 
    "is_vector_normed" = "false",  // whether the vectors are normalized or not
  )
  COMMENT 'embedding vector index'
) ENGINE=OLAP 
PRIMARY KEY(`id`) 
DISTRIBUTED BY HASH(`id`) BUCKETS 1 
PROPERTIES (
"replication_num" = "3", 
"enable_persistent_index" = "false", 
"replicated_storage" = "false", 
"compression" = "LZ4" 
);
```

ANNS:
```sql
select * from vector_table
   ORDER BY approx_l2_distance('0.1, 0.2, 0.3, ...., 0.1', question_embedding) 
   limit 10; 
```

Filtered ANNS:
```sql
select * from vector_table
   where source = "wiki"
   ORDER BY approx_l2_distance('0.1, 0.2, 0.3, ...., 0.1', question_embedding) 
   limit 10; 
```

Range search:
```sql
select * from vector_table 
    where approx_l2_distance(0.1, 0.2, 0.3, ...., 0.1', question_embedding) < 100; 
```

Filtered range search:
```sql
select * from vector_table 
    where source = "wiki" 
    and approx_l2_distance(0.1, 0.2, 0.3, ...., 0.1', question_embedding) < 100; 
```

### PR DESCRIPTION
**Pull request by dujijun007**
## What I'm doing:

Fixes #46678

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: thirdparty/build-thirdparty.sh [hunk 1]
```diff
@@ -1319,6 +1319,17 @@ build_simdutf() {
...
 
+# tenann
+build_tenann() {
+    check_if_source_exist $TENANN_SOURCE
+    rm -rf $TP_INSTALL_DIR/include/tenann
+    rm -rf $TP_INSTALL_DIR/lib/libtenann-bundle.a
+    rm -rf $TP_INSTALL_DIR/lib/libtenann-bundle-avx2.a
+    cp -r $TP_SOURCE_DIR/$TENANN_SOURCE/include/tenann $TP_INSTALL_DIR/include/tenann
+    cp -r $TP_SOURCE_DIR/$TENANN_SOURCE/lib/libtenann-bundle.a $TP_INSTALL_DIR/lib/
+    cp -r $TP_SOURCE_DIR/$TENANN_SOURCE/lib/libtenann-bundle-avx2.a $TP_INSTALL_DIR/lib/
```
Comment (kevincai) [inline thirdparty/build-thirdparty.sh:1330]: this prebuilt libraries won't work for multiple platforms, remember that we now support both x86 and arm platform.

**Review by kevincai — CHANGES_REQUESTED** [review]
need to fix multi-arch build problem.

File: thirdparty/build-thirdparty.sh (hunk 1 above)
Comment (dujijun007) [reply thirdparty/build-thirdparty.sh:1330]: you are right, we will build an additional libraries for the ARM platform.

**Review by decster — DISMISSED** [review]

File: be/CMakeLists.txt [hunk 2]
```diff
@@ -97,6 +97,16 @@ option(USE_AVX2 "Build with AVX2 instruction" ON)
 
 option(USE_AVX512 "Build with AVX512f/AVX512BW instruction" OFF)
 
+# Only support tenann on x86
+if ("${CMAKE_BUILD_TARGET_ARCH}" STREQUAL "x86" OR "${CMAKE_BUILD_TARGET_ARCH}" STREQUAL "x86_64")
```
Comment (kevincai) [inline be/CMakeLists.txt:101]: is it possible to turn off TENANN from build script?

File: thirdparty/build-thirdparty.sh [hunk 3]
```diff
@@ -1412,6 +1423,7 @@ build_fiu
 build_llvm
 build_clucene
 build_simdutf
+build_tenann
```
Comment (kevincai) [inline thirdparty/build-thirdparty.sh:1426]: make this step conditional. only do the build for x86 platform. check line 1428~1431

Comment (dujijun007) [reply thirdparty/build-thirdparty.sh:1426]: Got it

**Review by kevincai — DISMISSED** [review]

**Review by kevincai — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

File: thirdparty/build-thirdparty.sh (hunk 1 above)
Comment (JuanJuanQingQuan) [reply thirdparty/build-thirdparty.sh:1330]: Does the x86 platform contain Mac Intel? I test to build the tenann source code in mac intel, It occurs some error

Comment (kevincai) [reply thirdparty/build-thirdparty.sh:1330]: it's known to be broken on Mac, starrocks can be built natively on Mac for now.

**Pull request merged**